lib/kdtree.cpp

```diff
@@ -387,21 +533,60 @@ const KDTree::OptionalId KDTreeIntersection::intersect(const Ray& ray, float& r,
     return res;
 }
 
+// const KDTree::OptionalId
+// KDTreeIntersection::intersect(const KDTree::TriangleId* ids, uint32_t
+// num_tris,
+//                               const Ray& ray, float& min_r, float& min_s,
+//                               float& min_t) {
+//     min_r = std::numeric_limits<float>::max();
+//     OptionalId res;
+//     float r, s, t;
+//     for (uint32_t i = 0; i != num_tris; ++i) {
+//         const auto& tri = tree_->tris_[ids[i]];
+//         bool intersects = intersect_ray_triangle(ray, tri, r, s, t);
+//         if (intersects && r < min_r) {
+//             min_r = r;
+//             min_s = s;
+//             min_t = t;
+//             res = OptionalId{ids[i]};
+//         }
+//     }
+
+//     return res;
+// }
+
 const KDTree::OptionalId
-KDTreeIntersection::intersect(const KDTree::TriangleId* ids, uint32_t num_tris,
-                              const Ray& ray, float& min_r, float& min_s,
-                              float& min_t) {
+KDTreeIntersection::intersect(uint32_t node_index, const Ray& ray, float& min_r,
+                              float& min_s, float& min_t) {
     min_r = std::numeric_limits<float>::max();
     OptionalId res;
-    float r, s, t;
-    for (uint32_t i = 0; i != num_tris; ++i) {
-        const auto& tri = tree_->tris_[ids[i]];
+
+    auto intersect = [&](uint32_t triangle_id) {
```

Comment:
Did not want to repeat myself. So this is a small snippet. I don't think we should put it into a private function.